prediction_market_agent/agents/microchain_agent/deploy.py

```diff
@@ -97,64 +105,105 @@ def run_general_agent(
         # Save formatted system prompt
         initial_formatted_system_prompt = agent.system_prompt
 
-        try:
-            agent.run(self.n_iterations)
-        except Exception as e:
-            logger.error(e)
-            raise e
-        finally:
-            if goal_manager:
-                goal = check_not_none(goal)
-                goal_evaluation = goal_manager.evaluate_goal_progress(
-                    goal=goal,
-                    chat_history=ChatHistory.from_list_of_dicts(agent.history),
-                )
-                goal_manager.save_evaluated_goal(
-                    goal=goal,
-                    evaluation=goal_evaluation,
+        iteration = 0
+        while self.max_iterations is None or iteration < self.max_iterations:
+            starting_history_length = len(agent.history)
+            try:
+                # After the first iteration, resume=True to not re-initialize the agent.
+                agent.run(iterations=1, resume=iteration > 0)
+            except Exception as e:
+                logger.error(f"Error while running microchain agent: {e}")
+                raise e
+            finally:
+                # Save the agent's history to the long-term memory after every iteration to keep users updated.
+                save_agent_history(
+                    agent=agent,
+                    long_term_memory=long_term_memory,
+                    initial_system_prompt=initial_formatted_system_prompt,
+                    # Because the agent is running in a while cycle, always save into database only what's new, to not duplicate entries.
+                    save_last_n=len(agent.history) - starting_history_length,
                 )
-                agent.history.append(
-                    ChatMessage(
-                        role="user",
-                        content=str(f"# Goal evaluation\n{goal_evaluation}"),
-                    ).model_dump()
+                if agent.system_prompt != initial_formatted_system_prompt:
+                    prompt_handler.save_prompt(get_editable_prompt_from_agent(agent))
+            iteration += 1
+            logger.info(f"{self.__class__.__name__} iteration {iteration} completed.")
+            if self.sleep_between_iterations:
+                logger.info(
+                    f"{self.__class__.__name__} sleeping for {self.sleep_between_iterations} seconds."
                 )
+                time.sleep(self.sleep_between_iterations)
 
+        if goal_manager:
```

Comment:
Can we separate the logic between `goal_manager` and other cases? This method has ~100 lines and should be shortened.

Reply:
sure makes sense